modules/core/src/main/scala/doobie/util/read.scala

```diff
@@ -42,67 +43,133 @@ and similarly with Get:
+  def apply[A](implicit ev: Read[A]): Read[A] = ev
 
-  def derived[A](implicit ev: MkRead[A]): Read[A] = ev
+  def derived[A](implicit
+      @implicitNotFound(
+        "Cannot derive Read instance. Please check that each field in the case class has a Read instance or can derive one")
+      ev: Derived[MkRead[A]]
+  ): Read[A] = ev.instance.underlying
 
-  trait Auto {
-    implicit def deriveRead[A](implicit ev: MkRead[A]): Read[A] = ev
-  }
+  trait Auto extends MkReadInstances
 
   implicit val ReadApply: Applicative[Read] =
     new Applicative[Read] {
       def ap[A, B](ff: Read[A => B])(fa: Read[A]): Read[B] = fa.ap(ff)
-      def pure[A](x: A): Read[A] = new Read(Nil, (_, _) => x) {}
+      def pure[A](x: A): Read[A] = Read.unit.map(_ => x)
       override def map[A, B](fa: Read[A])(f: A => B): Read[B] = fa.map(f)
     }
 
-  implicit val unit: Read[Unit] =
-    Read(Nil, (_, _) => ())
+  implicit val unit: Read[Unit] = new Read[Unit] {
+    override def unsafeGet(rs: ResultSet, startIdx: Int): Unit = {
+      () // Does not read anything from ResultSet
+    }
+    override def gets: List[(Get[?], NullabilityKnown)] = List.empty
+    override def toOpt: Read[Option[Unit]] = optionUnit
+    override def length: Int = 0
+  }
+
+  implicit val optionUnit: Read[Option[Unit]] = unit.map(_ => Some(()))
+
+  implicit def fromReadOption[A](implicit read: Read[A]): Read[Option[A]] = read.toOpt
+
+  /** Simple instance wrapping a Get. i.e. single column non-null value */
+  class Single[A](get: Get[A]) extends Read[A] {
+    def unsafeGet(rs: ResultSet, startIdx: Int): A =
+      get.unsafeGetNonNullable(rs, startIdx)
+
+    override def toOpt: Read[Option[A]] = new SingleOpt(get)
+
+    override def gets: List[(Get[?], NullabilityKnown)] = List(get -> NoNulls)
+
+    override val length: Int = 1
+
+  }
+
+  /** Simple instance wrapping a Get. i.e. single column nullable value */
+  class SingleOpt[A](get: Get[A]) extends Read[Option[A]] {
+    def unsafeGet(rs: ResultSet, startIdx: Int): Option[A] =
+      get.unsafeGetNullable(rs, startIdx)
+
+    override def toOpt: Read[Option[Option[A]]] = new Transform[Option[Option[A]], Option[A]](this, a => Some(a))
+    override def gets: List[(Get[?], NullabilityKnown)] = List(get -> Nullability.Nullable)
+
+    override val length: Int = 1
+  }
+
+  class Transform[A, From](underlyingRead: Read[From], f: From => A) extends Read[A] {
+    override def unsafeGet(rs: ResultSet, startIdx: Int): A = f(underlyingRead.unsafeGet(rs, startIdx))
+    override def gets: List[(Get[?], NullabilityKnown)] = underlyingRead.gets
+    override def toOpt: Read[Option[A]] =
+      new Transform[Option[A], Option[From]](underlyingRead.toOpt, opt => opt.map(f))
+    override lazy val length: Int = underlyingRead.length
+  }
+
+  /** A Read instance consists of multiple underlying Read instances */
+  class Composite[A, S0, S1](read0: Read[S0], read1: Read[S1], f: (S0, S1) => A) extends Read[A] {
```

Comment:
After some benchmarking, I have reworked `Composite` to be structured more like `HList`s rather than `List[Read[Any]]`. 
With `List[Read[Any]]` my understanding is that JVM cannot effectively perform inlining because due to the dynamic length/type nature of the list.
But with a HList-like structure (e.g. `Composite(Read[Int], Composite(Read[String], Read[Int]))`) the call graph and types are "constant" which I think made the JIT happier

Using benchmark from #2139, the current implementation has basically no penalty between Read[A] and Read[Option[A]], whereas the previous implementation has around a 50% performance penalty.

```
[info] Benchmark                                Mode  Cnt        Score        Error  Units
[info] LargeRow.autoDerivedComplex             thrpt    5   956584.142 ± 105909.280  ops/s
[info] LargeRow.autoDerivedComplexOpt          thrpt    5   925193.390 ± 240656.559  ops/s
[info] LargeRow.rawJdbcComplex                 thrpt    5  1390634.422 ± 195647.819  ops/s
[info] LargeRow.rawJdbcTuple                   thrpt    5  1341330.033 ± 248449.784  ops/s
[info] LargeRow.semiautoDerivedComplex         thrpt    5   957341.195 ± 155499.630  ops/s
[info] LargeRow.semiautoDerivedComplexOpt      thrpt    5   917308.756 ± 181313.503  ops/s
[info] LargeRow.tuple                          thrpt    5  1068742.195 ± 208005.075  ops/s
[info] LargeRow.tupleOpt                       thrpt    5  1072509.030 ± 177644.605  ops/s
```